### PR TITLE
Refactor: branch by abstraction for Pipeline Launcher

### PR DESCRIPTION
**Pull request by henry-lp**
Abstracting the pipeline for (#1061)

The file diff is a bit unclear on `Launcher` due to many line deleted, therefore viewing the file directly is advised :) .

### CODE REVIEW

**Comment by henry-lp** [comment]
Green. It's review-ready :) @monperrus

File: src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipeline/NoopListener.java [hunk 1]
```diff
@@ -9,13 +9,13 @@
  */
 public class NoopListener implements Listener{
     private static final Logger LOGGER = LoggerFactory.getLogger(NoopListener.class);
-    private Launcher launcher;
+    private LauncherAPI launcher;
```
Comment (monperrus) [inline src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipeline/NoopListener.java]: Concrete-to-interface, that's one of the best refactoring ever!

File: src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipelineb/Launcher.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package fr.inria.spirals.repairnator.pipelineb;
+
+import fr.inria.spirals.repairnator.pipeline.LauncherAPI;
+import fr.inria.spirals.repairnator.config.RepairnatorConfig;
+import fr.inria.spirals.repairnator.process.inspectors.ProjectInspector;
+import fr.inria.spirals.repairnator.notifier.PatchNotifier;
+
+import com.martiansoftware.jsap.JSAP;
+
+public class Launcher implements LauncherAPI{
```
Comment (monperrus) [inline src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipelineb/Launcher.java]: we should not create a new package.


why do we need this class? remove?

File: src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipeline/NoopListener.java (hunk 1 above)
Comment (henry-lp) [reply src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipeline/NoopListener.java]: Dependency inversion principle :)

File: src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipelineb/Launcher.java (hunk 2 above)
Comment (henry-lp) [reply src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipelineb/Launcher.java]: It's the new Launcher . I changed its name to `BranchLauncher` and it's now in the same package :) .

File: src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipeline/Launcher.java [hunk 3]
```diff
@@ -1,573 +1,70 @@
...
-import java.util.Arrays;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Optional;
-import java.util.Properties;
 
-/**
- * This class is the main entry point for the repairnator pipeline.
- */
 public class Launcher {
```
[the remark this answers is not in the document]
Comment (henry-lp) [reply src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipeline/Launcher.java]: done :) .

File: src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipeline/LauncherAPI.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package fr.inria.spirals.repairnator.pipeline;
+
+import fr.inria.spirals.repairnator.config.RepairnatorConfig;
+import fr.inria.spirals.repairnator.process.inspectors.ProjectInspector;
+import fr.inria.spirals.repairnator.notifier.PatchNotifier;
+
+import com.martiansoftware.jsap.JSAP;
+import com.martiansoftware.jsap.JSAPException;
+
+public interface LauncherAPI {
```
[the remark this answers is not in the document]
Comment (henry-lp) [reply src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/pipeline/LauncherAPI.java]: done :) .

**Comment by monperrus** [comment]
Thanks a lot Henry.

**Pull request merged**